data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.datanode.initializers;
+
+import io.jsonwebtoken.Jwt;
+import io.jsonwebtoken.JwtParser;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.crypto.spec.SecretKeySpec;
+import javax.inject.Named;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.Key;
+import java.util.Collections;
+import java.util.Optional;
+
+/**
+ * This is an authorization filter that  first try to verify presence and validity of a JWT token. If there is no
+ * bearer token available, it will fallback to basic auth (or whatever filter is configured as fallback).
+ * Allowing both auth methods allows easy access directly from CLI or browser and machine-machine communication from the graylog server.
+ */
+public class DatanodeAuthFilter implements ContainerRequestFilter {
```

Comment:
**We've been talking with Tomas a little bit about it on Slack.
Tomas asked me to put my comments here. It is not a strict request, rather an idea that could be considered.**

1. The thing I'd do is to separate this stuff to 2 separate classes, moving JWT-related stuff to some new class and making `DatanodeAuthFilter` responsible for trying some auth method and using other as a fallback.
But the concrete method - JWT - I'd have in another class, passed to `DatanodeAuthFilter` via constructor.

2. I would have an interface, let's say `AuthVerification` and its implemenation `JwtAuthVerification`. In the test I'd mock `AuthVerification` without referring to any JWT-like details. I'd test `JwtAuthVerification` in a separate unit test.
You have a "classicist" test - like mini-integration test, which is enforced by the different coding style.

3. That's generally another example of slight differences of our coding styles. I try more to keep "single responsibility" from the start (probably I fail it often as well...), you would rather sacrifice this principle at the early stage, just not to have too much code that you deem unnecessary at that point of time.

Reply:
Thanks @luk-kaminski , I refactored the mentioned parts of the code and both the code and tests are now cleaner and more readable.